cmd/snap/cmd_components.go

```diff
@@ -0,0 +1,133 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2024 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package main
+
+import (
+	"fmt"
+	"sort"
+
+	"github.com/jessevdk/go-flags"
+
+	"github.com/snapcore/snapd/client"
+	"github.com/snapcore/snapd/i18n"
+	"github.com/snapcore/snapd/snap/naming"
+)
+
+var shortComponentsHelp = i18n.G("List available and installed components for installed snaps")
+var longComponentsHelp = i18n.G(`
+The components command displays a summary of the components that are installed
+and available for the set of currently installed snaps.
+
+Components for specific installed snaps can be queried by providing snap names
+as positional arguments.
+`)
+
+type cmdComponents struct {
+	clientMixin
+	Positional struct {
+		Snaps []installedSnapName `positional-arg-name:"<snap>"`
+	} `positional-args:"yes"`
+}
+
+func init() {
+	addCommand("components",
+		shortComponentsHelp,
+		longComponentsHelp,
+		func() flags.Commander { return &cmdComponents{} },
+		nil,
+		[]argDesc{{
+			name: i18n.G("<snap>"),
+			desc: i18n.G("Snaps to consider when listing available and installed components."),
+		}},
+	)
+}
+
+func (x *cmdComponents) Execute(args []string) error {
+	if len(args) > 0 {
+		return ErrExtraArgs
+	}
```

Comment:
I don't think we can ever hit this? Since the last positional argument is a slice. Same with the snap list command I suppose